### PR TITLE
Test path-based routing via HTTPS and WebSockets

### PR DESCRIPTION
**Pull request by elyscape**
As mentioned [in this comment](https://github.com/openshift/origin/pull/8490#issuecomment-214884895), the router's path-based integration tests only really go in-depth via HTTP. Issues can manifest on HTTP without manifesting on HTTPS and vice versa, as well as on WebSockets (for which there is no path-based route testing). As such, this pull request adds integration testing for all those scenarios.

### CODE REVIEW

**Comment by elyscape** [comment]
@ramr, please take a look.

File: test/integration/router_test.go [hunk 1]
```diff
@@ -447,6 +469,13 @@ func TestRouterPathSpecificity(t *testing.T) {
 				To: kapi.ObjectReference{
 					Name: "myService",
 				},
+				routeTLS: &routeapi.TLSConfig{
```
Comment (ramr) [inline test/integration/router_test.go]: field is named `TLS` not `routeTLS`.

File: test/integration/router_test.go [hunk 2]
```diff
@@ -457,16 +486,18 @@ func TestRouterPathSpecificity(t *testing.T) {
...
-	if _, err := getRoute(routeAddress, "www.example.com", "http", nil, ""); err != ErrUnavailable {
-		t.Fatalf("unexpected response: %q", err)
-	}
-	//ensure you can curl path with port in Host header
-	if err := waitForRoute(routeTestAddress, "www.example.com:80", "http", nil, tr.HelloPodPath); err != nil {
-		t.Fatalf("unexpected response: %q", err)
+	for _, proto := range protocols {
+		//ensure you can curl path but not main host
+		if err := waitForRoute(routeTestAddress, "www.example.com", proto.name, nil, tr.HelloPodPath); err != nil {
+			t.Fatalf("unexpected response: %q", err)
```
Comment (ramr) [inline test/integration/router_test.go]: Can you please also add which `protocol` specific test failed in the error messages.

File: test/integration/router_test.go [hunk 3]
```diff
@@ -457,16 +486,18 @@ func TestRouterPathSpecificity(t *testing.T) {
...
+		}
+		if _, err := getRoute(routeAddress, "www.example.com", proto.name, nil, ""); err != ErrUnavailable {
+			t.Fatalf("unexpected response: %q", err)
+		}
+		//ensure you can curl path with port in Host header
+		if err := waitForRoute(routeTestAddress, "www.example.com:" + proto.port, proto.name, nil, tr.HelloPodPath); err != nil {
+			t.Fatalf("unexpected response: %q", err)
```
Comment (ramr) [inline test/integration/router_test.go]: similar comment as above re: the error message text.

File: test/integration/router_test.go [hunk 4]
```diff
@@ -457,16 +486,18 @@ func TestRouterPathSpecificity(t *testing.T) {
...
-	//ensure you can curl path with port in Host header
-	if err := waitForRoute(routeTestAddress, "www.example.com:80", "http", nil, tr.HelloPodPath); err != nil {
-		t.Fatalf("unexpected response: %q", err)
+	for _, proto := range protocols {
+		//ensure you can curl path but not main host
+		if err := waitForRoute(routeTestAddress, "www.example.com", proto.name, nil, tr.HelloPodPath); err != nil {
+			t.Fatalf("unexpected response: %q", err)
+		}
+		if _, err := getRoute(routeAddress, "www.example.com", proto.name, nil, ""); err != ErrUnavailable {
+			t.Fatalf("unexpected response: %q", err)
```
Comment (ramr) [inline test/integration/router_test.go]: similar comment as above re: the error message text.

File: test/integration/router_test.go [hunk 5]
```diff
@@ -494,17 +525,26 @@ func TestRouterPathSpecificity(t *testing.T) {
 				To: kapi.ObjectReference{
 					Name: "altService",
 				},
+				routeTLS: &routeapi.TLSConfig{
```
Comment (ramr) [inline test/integration/router_test.go]: Change to `TLS` instead of `routeTLS`.

File: test/integration/router_test.go [hunk 6]
```diff
@@ -494,17 +525,26 @@ func TestRouterPathSpecificity(t *testing.T) {
...
-		t.Fatalf("unexpected response: %q", err)
-	}
-	if err := waitForRoute(routeTestAddress, "www.example.com:80", "http", nil, tr.HelloPodPath); err != nil {
-		t.Fatalf("unexpected response: %q", err)
+	for _, proto := range protocols {
+		if err := waitForRoute(routeTestAddress, "www.example.com", proto.name, nil, tr.HelloPodPath); err != nil {
+			t.Fatalf("unexpected response: %q", err)
+		}
+		if err := waitForRoute(routeTestAddress, "www.example.com:" + proto.port, proto.name, nil, tr.HelloPodPath); err != nil {
+			t.Fatalf("unexpected response: %q", err)
```
Comment (ramr) [inline test/integration/router_test.go]: same as previous comment re: adding `protocol` (`proto.name`)  information to the failure message.

File: test/integration/router_test.go [hunk 7]
```diff
@@ -494,17 +525,26 @@ func TestRouterPathSpecificity(t *testing.T) {
...
 	fakeMasterAndPod.RouteChannel <- eventString(routeEvent)
 
-	if err := waitForRoute(routeTestAddress, "www.example.com", "http", nil, tr.HelloPodPath); err != nil {
-		t.Fatalf("unexpected response: %q", err)
-	}
-	if err := waitForRoute(routeTestAddress, "www.example.com:80", "http", nil, tr.HelloPodPath); err != nil {
-		t.Fatalf("unexpected response: %q", err)
+	for _, proto := range protocols {
+		if err := waitForRoute(routeTestAddress, "www.example.com", proto.name, nil, tr.HelloPodPath); err != nil {
+			t.Fatalf("unexpected response: %q", err)
```
Comment (ramr) [inline test/integration/router_test.go]: same as previous comment re: adding `protocol` (`proto.name`)  information to the failure message.

File: test/integration/router_test.go [hunk 8]
```diff
@@ -521,20 +561,29 @@ func TestRouterPathSpecificity(t *testing.T) {
 				To: kapi.ObjectReference{
 					Name: "myService",
 				},
+				routeTLS: &routeapi.TLSConfig{
```
Comment (ramr) [inline test/integration/router_test.go]: change to `TLS` in lieu of `routeTLS`.

File: test/integration/router_test.go [hunk 9]
```diff
@@ -521,20 +561,29 @@ func TestRouterPathSpecificity(t *testing.T) {
...
+		//ensure you can curl path and host
+		if err := waitForRoute(routeTestAddress, "www.example.com", proto.name, nil, tr.HelloPodPath); err != nil {
+			t.Fatalf("unexpected response: %q", err)
+		}
+		if err := waitForRoute(routeAddress, "www.example.com", proto.name, nil, tr.HelloPod); err != nil {
+			t.Fatalf("unexpected response: %q", err)
+		}
+		if err := waitForRoute(routeTestAddress, "www.example.com:" + proto.port, proto.name, nil, tr.HelloPodPath); err != nil {
+			t.Fatalf("unexpected response: %q", err)
+		}
```
Comment (ramr) [inline test/integration/router_test.go]: same comments as earlier re: adding `protocol` information to the failure messages (the 3 `t.Fatalf`s in the for loop).

File: test/integration/router_test.go [hunk 10]
```diff
@@ -551,6 +600,13 @@ func TestRouterPathSpecificity(t *testing.T) {
 				To: kapi.ObjectReference{
 					Name: "myService",
 				},
+				routeTLS: &routeapi.TLSConfig{
```
Comment (ramr) [inline test/integration/router_test.go]: change to `TLS` (not `routeTLS`).

File: test/integration/router_test.go [hunk 11]
```diff
@@ -562,14 +618,16 @@ func TestRouterPathSpecificity(t *testing.T) {
...
+	for _, proto := range protocols {
+		if err := waitForRoute(routeTestAddress, "www.example.com", proto.name, nil, tr.HelloPodPath); err != nil {
+			t.Fatalf("unexpected response: %q", err)
+		}
+		if err := waitForRoute(routeAddress, "www.example.com", proto.name, nil, tr.HelloPod); err != nil {
+			t.Fatalf("unexpected response: %q", err)
+		}
+		if err := waitForRoute(routeTestAddress, "www.example.com:" + proto.port, proto.name, nil, tr.HelloPodPath); err != nil {
+			t.Fatalf("unexpected response: %q", err)
+		}
```
Comment (ramr) [inline test/integration/router_test.go]: same comments as earlier re: adding protocol information to the failure messages (the 3 `t.Fatalf`s in the for loop).

File: test/integration/router_test.go [hunk 12]
```diff
@@ -586,19 +644,28 @@ func TestRouterPathSpecificity(t *testing.T) {
 				To: kapi.ObjectReference{
 					Name: "altService",
 				},
+				routeTLS: &routeapi.TLSConfig{
```
Comment (ramr) [inline test/integration/router_test.go]: correct field name: `TLS`.

File: test/integration/router_test.go [hunk 13]
```diff
@@ -586,19 +644,28 @@ func TestRouterPathSpecificity(t *testing.T) {
...
+	for _, proto := range protocols {
+		if err := waitForRoute(routeTestAddress, "www.example.com", proto.name, nil, tr.HelloPodPath); err != nil {
+			t.Fatalf("unexpected response: %q", err)
+		}
+		if err := waitForRoute(routeAddress, "www.example.com", proto.name, nil, tr.HelloPod); err != nil {
+			t.Fatalf("unexpected response: %q", err)
+		}
+		if err := waitForRoute(routeTestAddress, "www.example.com:" + proto.port, proto.name, nil, tr.HelloPodPath); err != nil {
+			t.Fatalf("unexpected response: %q", err)
+		}
```
Comment (ramr) [inline test/integration/router_test.go]: same comments as earlier re: adding protocol information to the failure messages (the 3 `t.Fatalf`s in the for loop).

File: test/integration/router_test.go [hunk 14]
```diff
@@ -615,19 +682,28 @@ func TestRouterPathSpecificity(t *testing.T) {
 				To: kapi.ObjectReference{
 					Name: "altService",
 				},
+				routeTLS: &routeapi.TLSConfig{
```
Comment (ramr) [inline test/integration/router_test.go]: `TLS`

File: test/integration/router_test.go [hunk 15]
```diff
@@ -643,6 +719,13 @@ func TestRouterPathSpecificity(t *testing.T) {
 				To: kapi.ObjectReference{
 					Name: "myService",
 				},
+				routeTLS: &routeapi.TLSConfig{
```
Comment (ramr) [inline test/integration/router_test.go]: correct field name: `TLS`

File: test/integration/router_test.go [hunk 16]
```diff
@@ -615,19 +682,28 @@ func TestRouterPathSpecificity(t *testing.T) {
...
+	for _, proto := range protocols {
+		if err := waitForRoute(routeTestAddress, "www.example.com", proto.name, nil, tr.HelloPodAlternate); err != nil {
+			t.Fatalf("unexpected response: %q", err)
+		}
+		if err := waitForRoute(routeAddress, "www.example.com", proto.name, nil, tr.HelloPodAlternate); err != nil {
+			t.Fatalf("unexpected response: %q", err)
+		}
+		if err := waitForRoute(routeTestAddress, "www.example.com:" + proto.port, proto.name, nil, tr.HelloPodAlternate); err != nil {
+			t.Fatalf("unexpected response: %q", err)
+		}
```
Comment (ramr) [inline test/integration/router_test.go]: same comments as earlier re: adding protocol information to the failure messages (the 3 `t.Fatalf`s in the for loop).

**Comment by elyscape** [comment]
@ramr: Comments addressed.

**Comment by elyscape** [comment]
Fixed a formatting issue that `gofmt` didn't like.

**Comment by ramr** [comment]
LGTM

**Comment by ramr** [comment]
[test]

**Comment by elyscape** [comment]
Looks like this failed due to an issue [documented here](https://github.com/openshift/origin/blob/v1.3.0-alpha.0/test/integration/router_test.go#L296-L310). In short, the Go WebSockets client doesn't support customizing the `Host` header, which causes issues when trying to test routes with domains. While this can be worked around during the regular testing by setting the host to `0.0.0.0`, I'm not sure how well it would work specifically during route path testing. Thoughts, @ramr? I wonder if we might be better served by finding a way to ensure that the domain name always resolves to the router host's IP address.

**Comment by elyscape** [comment]
So, setting up dnsmasq locally to return `127.0.0.1` for `www.example.com` worked, but then lead to another problem: the test HTTP server only listens for WebSocket requests that go to `/echo`. It's going to need some reworking to:

1. support WebSocket requests on arbitrary paths and
2. support both WebSocket and regular HTTP requests on the same path

**Comment by elyscape** [comment]
Okay, I reworked the test HTTP server and everything checks out. The trick now is to figure out how to either get dnsmasq set up in the test infrastructure or figure out some other solution. I'm also looking into getting dnsmasq configured in the Vagrant box.

**Comment by elyscape** [comment]
Actually, since it looks like the test infrastructure uses Vagrant to configure the test systems, if I get the Vagrant box configured then everything should follow from there. Some issues, though:

0. How is F5 BIG-IP routing tested? Is it done inside a Vagrant box or in a separate, pre-configured cluster? If `www.example.com` always resolves to the local machine when using Vagrant, it won't be possible to test F5 BIG-IP routing.
0. How should multi-node and docker-in-docker clusters be handled?
0. What about users developing/testing locally on their host machine?

**Comment by ramr** [comment]
@elyscape in response to your questions:
1. F5 BIGIP testing is done on an external environment - there's 2 broad sets of tests (the unit/integration tests using a mocked F5 service) and the tests with an actual F5 BIGIP instance that QE runs separately. The external F5 integration tests requires some setup today to "bridge" F5 and the OpenShift cluster via a ramp node (instructions at: https://docs.openshift.org/latest/install_config/routing_from_edge_lb.html and create an F5 route synchronizer are at: https://docs.openshift.org/latest/install_config/install/deploy_router.html#deploying-the-f5-router). 

Hmm, am not sure if adding a local resolution to `www.example.com` is a good fix (we should just do it in the tests rather than need to change the system/vagrant config). I would try to get a way to do this in the tests ("scoped" to the process the tests run in).
On 2. and 3. see above - if we do it without needing a change on the env (hosts/dns server etc), then this point becomes moot.

**Comment by elyscape** [comment]
@ramr Unfortunately, as far as I can tell, there's no way to modify the Go DNS resolution process beyond selecting whether to use the pure Go DNS resolver or call out to C library functions `getaddrinfo`, `getnameinfo`, and similar ([reference][go-vs-crt]), and that doesn't work on Plan 9 or Windows (neither of which are supported by Origin, so whatever). Forcing the cgo resolver by setting `GODEBUG=netdns=cgo` in the environment should allow us to configure domain names in the `/etc/hosts` file, but then we're back to needing to modify system configuration for the test. It _might_ be possible to set up something using [the `HOSTALIASES` environment variable][hostaliases], but experimentation seems to indicate that, after the mapping in the `HOSTALIASES` file is applied, the new hostname is used to query DNS directly without a step to check the `/etc/hosts` file, so trying to map `example localhost.localdomain` fails. We'd need to set up some hostname on the internet that resolved to `127.0.0.1` for that to work. Then again, it looks like `localdomain.me` [resolves to `127.0.0.1`][localdomain.me], so we could use that as an alias target.

Thoughts?

[go-vs-crt]: https://golang.org/pkg/net/#hdr-Name_Resolution
[hostaliases]: http://www.unix.com/man-page/centos/7/hostname/
[localdomain.me]: http://readme.localtest.me/

**Comment by liggitt** [comment]
typically for tests like that, I've just used a custom dialer that always dials the IP of the router, ignoring the passed hostname

**Comment by liggitt** [comment]
see https://github.com/openshift/origin/blob/master/test/integration/sni_test.go#L210-L213 for an example

instead of calling `websocket.DialConfig`, you'd do the dial yourself to get a connection, then call `websocket.NewClient(config, conn)`

**Comment by elyscape** [comment]
Interesting. I'll take a stab at it.

**Comment by elyscape** [comment]
@liggit Thanks for the tip. I have implemented it here. It should solve all my earlier concerns, with one possible exception: does the router for multi-node and docker-in-docker clusters run on the machine executing the tests? It looks like it does, in which case this should work fine, but I just wanted to verify.

File: test/integration/router_test.go [hunk 17]
```diff
@@ -1048,16 +1129,25 @@ func getRoute(routerUrl string, hostName string, protocol string, headers map[st
...
+			return "", err
+		}
+		ws, err := websocket.NewClient(wsConfig, conn)
 		if err != nil {
-			if derr, ok := err.(*websocket.DialError); ok {
-				if derr.Err == websocket.ErrBadStatus {
-					// a better websocket library would know the difference here
-					return "", ErrUnavailable
-				}
+			if err == websocket.ErrBadStatus {
```
Comment (elyscape) [inline test/integration/router_test.go]: This is the only part I'm particularly unsure about. Does this look right?

Comment (liggitt) [inline test/integration/router_test.go]: in createAndStartRouterContainer, can't we get the IP of the router container?

**Comment by liggitt** [comment]
> does the router for multi-node and docker-in-docker clusters run on the machine executing the tests? It looks like it does, in which case this should work fine, but I just wanted to verify.

in createAndStartRouterContainer, can't we get the IP of the router container? what were we doing previously, using 0.0.0.0?

File: test/integration/router_test.go (hunk 17 above)
Comment (liggitt) [inline test/integration/router_test.go]: the dial error case would have been encountered a few lines above, during the dial

Comment (liggitt) [inline test/integration/router_test.go]: mmm... maybe not

**Comment by elyscape** [comment]
> in createAndStartRouterContainer, can't we get the IP of the router container?
That should work, now that you mention it. Then again, it does map that port onto the host machine, so maybe not? I forget what happens in that situation; I'll need to try it out.
> what were we doing previously, using 0.0.0.0?
Yup.

**Comment by elyscape** [comment]
> in createAndStartRouterContainer, can't we get the IP of the router container?

Now that I think about it more, this isn't a good answer and won't always work. If the container is on the host running the test, using `127.0.0.1` would work fine. If the container is on some other host, though, the testing host won't be able to reach the container's IP address. What we need is some way to get the IP address of the docker host itself. I'm looking into that now.

**Comment by elyscape** [comment]
There appears to be no way to query the Docker API for the host's IP address, which makes sense; it could have several, and who knows which one would be useful to you. Additionally, either you already know it (because you're connecting via TCP and can deduce it from the `DOCKER_HOST` environment variable) or you're on the same machine and are connecting via a Unix socket and should figure it out yourself.

I'll look into extracting the information from the Docker API client object.

**Comment by elyscape** [comment]
Actually, it looks like we're already assuming that the router runs on the local system:

https://github.com/openshift/origin/blob/v1.3.0-alpha.0/test/integration/router_test.go#L421
https://github.com/openshift/origin/blob/v1.3.0-alpha.0/test/integration/router_test.go#L672
https://github.com/openshift/origin/blob/v1.3.0-alpha.0/test/integration/router_test.go#L799
https://github.com/openshift/origin/blob/v1.3.0-alpha.0/test/integration/router_test.go#L801
https://github.com/openshift/origin/blob/v1.3.0-alpha.0/test/integration/router_test.go#L853
https://github.com/openshift/origin/blob/v1.3.0-alpha.0/test/integration/router_test.go#L890

So that's a thing.

I'll still look into having it pull the information from the Docker API client object, but perhaps that should be a second pull request?

File: test/integration/router_test.go (hunk 17 above)
Comment (elyscape) [inline test/integration/router_test.go]: Okay. I'll put that back.

Comment (elyscape) [inline test/integration/router_test.go]: Putting this back causes it to break, saying that `websocket.ErrBadStatus` is not the expected value of `ErrUnavailable`. Looks like I had done the right thing.

**Comment by elyscape** [comment]
Updated to use [`net.SplitHostPort`][SplitHostPort] instead of manually splitting out the port with string operations.

[SplitHostPort]: https://golang.org/pkg/net/#SplitHostPort

**Comment by elyscape** [comment]
Okay, this should be good to merge now. Once that's done, I'll submit another pull request to support non-local Docker hosts. I have the changes ready and working locally.

**Comment by ramr** [comment]
[test]

**Comment by ramr** [comment]
@elyscape  thx  - the changes LGTM.

@liggitt  do you have comments? thx

**Comment by liggitt** [comment]
Nothing significant. I wondered about the hard coded 127.0.0.1 and whether we can use the custom dialers for everything, not just the default router image, but no real objections.

**Comment by elyscape** [comment]
> I wondered about the hard coded 127.0.0.1 and whether we can use the custom dialers for everything, not just the default router image, but no real objections.

We could have it read from some environment variable. What should I call it?

**Comment by liggitt** [comment]
I was just confused why we couldn't dial the `routerUrl` instead of hard-coding 127.0.0.1

**Comment by elyscape** [comment]
Good call. I'll try changing it to use the output of `getRouteAddress()`.

**Comment by elyscape** [comment]
@liggit That worked. Change made and pushed.

**Comment by liggitt** [comment]
Great. Go ahead and squash, no more comments from me

**Comment by elyscape** [comment]
@liggitt, @ramr: Squashed.

**Comment by ramr** [comment]
[test]

**Comment by ramr** [comment]
@liggitt tests passed, could you please merge - thx.

**Comment by liggitt** [comment]
[merge]

**Pull request merged**